libraries/AC_Fence/AC_Fence.cpp

```diff
@@ -876,12 +905,48 @@ void AC_Fence::record_breach(uint8_t fence_type)
     _breached_fences |= fence_type;
 }
 
-/// clear_breach - update breach bitmask, time and count
+/// record_margin_breach - update margin_breach bitmask, time and count
+void AC_Fence::record_margin_breach(uint8_t fence_type)
+{
+    // if we haven't already breached a margin limit, update the breach time
+    if (!(_breached_fence_margins & fence_type)) {
+        const uint32_t now = AP_HAL::millis();
+
+        // emit a message indicated we're newly-breached, but not too often
+        if (option_enabled(OPTIONS::MARGIN_BREACH)
+            && AP_HAL::timeout_expired(_last_margin_breach_notify_sent_ms, now, 1000U)) {
+            _last_margin_breach_notify_sent_ms = now;
+            print_fence_message("outside margin", _breached_fence_margins | fence_type);
+        }
+    }
+
+    // update bitmask
+    _breached_fence_margins |= fence_type;
+}
+
+/// clear_breach - update breach bitmask
 void AC_Fence::clear_breach(uint8_t fence_type)
 {
     _breached_fences &= ~fence_type;
 }
 
+/// clear_margin_breach - update margin reach bitmask
+void AC_Fence::clear_margin_breach(uint8_t fence_type)
+{
+    if (_breached_fence_margins & fence_type) {
+        const uint32_t now = AP_HAL::millis();
+
+        // emit a message indicated we're newly-breached, but not too often
+        if (option_enabled(OPTIONS::MARGIN_BREACH)
+            && AP_HAL::timeout_expired(_last_margin_breach_notify_sent_ms, now, 1000U)) {
```

Comment:
parameterize "not too often"?

Reply:
I'm loathe to add another parameter unless its really required - we have many other repeating messages in warning situations that do not have a configurable repeat interval

Reply:
Fair enough, but this seems to be every second. (Do I have that wrong?) That seems pretty often to me.

Reply:
Yeah, open to a better value - although you obviously can turn it off. You only get the warning when you are inside the fence but outside the margin - its supposed to be naggy.

Reply:
As one of the requestors of this feature: I think it looks like the solution to our problem, which @andyp1per described perfectly above.
1 message per second is fine with me.

Reply:
@enok71 note that the polyfence support is very approximate. There is no margin support for polyfences so what I did was check a square box, centered on the current position with edges at the margin and see if any corner was outside the fence. Obviously lots of situations in which this is not quite accurate, but checking the polyfence is quite expensive and it wasn't clear to me if there was a better way without writing lots of code.

Reply:
Ok. I realize computing actual smallest distance to polygon exterior would be complicated.

Perhaps allowing two geofences would be a way to eliminate that? One warning-geofence and one failsafe action geofence.

Anyway, what you already have implemented is much better than nothing, so thanks once again.

Reply:
@timtuxworth pretty sure the avoid code already uses the margin - this is for the specific case where you are flying manually without avoid.

Reply:
Sure it does - but your calculation of the margin is different. I'm saying you should use the same calculation. Your approximation will give different results.